packages/contracts-bedrock/test/scripts/L2Genesis.t.sol

```diff
@@ -0,0 +1,232 @@
+            commands[2] = string.concat(
+                "DEPLOYMENT_CONTEXT=",
+                deploymentContext,
+                " forge script --chain-id ",
+                vm.toString(cfg.l2ChainID()),
+                " ./scripts/L2Genesis.s.sol:L2Genesis"
+            );
+            vm.ffi(commands);
+        }
+    }
+
+    /// @notice Iterates over every alloc parsed from `genesisPath`, and depending on
+    ///         the alloc address, runs specific checks based on whether the alloc is
+    ///         a precompile, predeploy proxy, or predeploy implementation. If the
+    ///         alloc address doesn't pass any of the checks to determine what it is,
+    ///         the function reverts with
+    ///         `string.concat("Unknown alloc: ", Strings.toHexString(allocs[i].addr)`.
+    function test_allocs() external {
+        Alloc[] memory allocs = _parseAllocs(genesisPath);
+
+        for (uint256 i; i < allocs.length; i++) {
+            uint160 numericAddress = uint160(allocs[i].addr);
+            if (numericAddress < L2GenesisHelpers.PRECOMPILE_COUNT) {
+                _checkPrecompile(allocs[i]);
+            } else if (_isProxyAddress(allocs[i].addr)) {
+                _checkProxy(allocs[i]);
+            } else if (_isImplementationAddress(allocs[i].addr)) {
+                _checkImplementation(allocs[i]);
+            } else if (_isDevAccount(allocs[i].addr)) {
+                _checkDevAccount(allocs[i]);
+            } else {
+                revert(string.concat("Unknown alloc: ", vm.toString(allocs[i].addr)));
+            }
+        }
+    }
+
+    /// @notice Runs checks against `_alloc` to determine if it's an expected precompile.
+    ///         The following should hold true for every precompile:
+    ///         1. The alloc should have a balance of `1`.
+    ///         2. The alloc should not have `code` set.
+    ///         3. The alloc should have a `nonce` of `0`.
+    ///         4. The alloc should not have any storage slots set.
+    function _checkPrecompile(Alloc memory _alloc) internal {
+        assertEq(_alloc.balance, hex"01");
+        assertEq(_alloc.code, hex"");
+        assertEq(_alloc.nonce, hex"00");
+        assertEq(_alloc.storageData.length, 0);
+    }
+
+    /// @notice Runs checks against `_alloc` to determine if it's an expected predeploy proxy.
+    ///         The following should hold true for every predeploy proxy:
+    ///         1. The alloc should have a balance of `0`.
+    ///         2. The alloc should have `code` set `Proxy.sol` deployed bytecode.
+    ///         3. The alloc should have a `nonce` of `0`.
+    ///         4. The alloc should two storage slots set:
+    ///            1. L2GenesisHelpers.PROXY_ADMIN_ADDRESS
+    ///            2. L2GenesisHelpers.PROXY_IMPLEMENTATION_ADDRESS
+    function _checkProxy(Alloc memory _alloc) internal {
+        assertEq(_alloc.balance, hex"00");
+        assertEq(_alloc.nonce, hex"00");
```

Comment:
> Where there is code, the nonce really should be 1 in the genesis. I think we should stick with it being 0 though as to not create issues between old and new states. I think this was an oversight in the original implementation

@tynes Do you think there's any downside to changing nonces to 1? I can't think of any consequences for either action (nonce 0 or 1) but it seems safer to avoid breaking EVM invariants and set nonce to 1 when there is code. Though to be fair there are contracts with nonce 0 since this wasn't a rule at L1 genesis (https://eips.ethereum.org/EIPS/eip-161), so either option should be safe. I don't feel too strongly in either direction here

Reply:
What it does impact directly are contracts that use `CREATE`. It makes me slightly concerned that we would miss something when doing an upgrade if the nonces were different in the contracts so I prefer to not change things now